app/models/category.rb

```diff
@@ -0,0 +1,68 @@
+class Category < ApplicationRecord
+  #ASSOCIATIONS
+
+  #Creating a self join on Category
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_category_id",
+            dependent: :destroy
+  belongs_to :parent_category, class_name: "Category", optional: true
+
+  #Other Associations
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_category_products, through: :sub_categories, source: :products
+
+  #Callbacks
+  before_destroy :check_sub_product_count, :check_product_count, prepend: true
+
+  #VALIDATIONS
+  validates :name, presence: true
+  validates :name, uniqueness: {scope: :parent_category_id},
+    if: -> { name }
+  validate :sub_category_has_one_level_nesting
+
+  private
+
+  #validation methods
+  def sub_category_has_one_level_nesting
```

Comment:
@avneetsng Add dirty objects check in validate to save redundant queries. 
Also optimize this method